js/capture.js

```
@@ -9,16 +9,59 @@ const capture = {
 
   addListeners() {
     // listen for each HTTP response
-    browser.webRequest.onResponseStarted.addListener(
-      (response) => this.sendThirdParty(response),
+    this.queue = [];
+    browser.webRequest.onResponseStarted.addListener((response) => {
+      const eventDetails = {
+        type: 'sendThirdParty',
+        data: response
+      };
+      this.queue.push(eventDetails);
+      this.processNextEvent();
+    },
       {urls: ['<all_urls>']});
     // listen for tab updates
     browser.tabs.onUpdated.addListener(
       (tabId, changeInfo, tab) => {
-        this.sendFirstParty(tabId, changeInfo, tab);
+        const eventDetails = {
+          type: 'sendFirstParty',
+          data: {
+            tabId: tabId,
+            changeInfo: changeInfo,
+            tab: tab
+          }
+        };
+        this.queue.push(eventDetails);
+        this.processNextEvent();
       });
   },
 
+  async processNextEvent(ignore = false) {
+    if (this.processingQueue && !ignore) {
+      return;
+    }
+    if (this.queue.length > 1) {
+      const nextEvent = this.queue.shift();
+      this.processingQueue = true;
+      switch (nextEvent.type) {
+        case 'sendFirstParty':
+          await this.sendFirstParty(
+            nextEvent.data.tabId,
+            nextEvent.data.changeInfo,
+            nextEvent.data.tab
+          );
+          break;
+        case 'sendThirdParty':
+          await this.sendThirdParty(nextEvent.data);
+          break;
+        default:
```

We should probably throw here rather than silently ignore.